hamilton/lifecycle/default.py

```diff
@@ -313,6 +316,109 @@ def run_after_node_execution(
             pdb.set_trace()
 
 
+class CacheAdapter(NodeExecutionHook, NodeExecutionMethod, GraphExecutionHook):
+    """Class to cache node results on disk with a key baed on the node code implementation and inputs.
+    Following runs with the same key can load node results and skip computation.
+
+    The cache `_nodes_history` entry returns an append-only list of results added to the cache.
+    e.g., the last value in the list `cache["_node_history"][node_name]` is the most recent cached node.
+
+    Notes:
+        - It uses the stdlib `shelve` module and the pickle format, which makes results dependent
+        on the Python version. Use materialization for persistent results
+        - There are no utility to manage cache size so you'll have to delete it periodically. Look
+        at the diskcache plugin for Hamilton `hamilton.plugins.h_diskcache` for better cache management.
+    """
+
+    nodes_history_key: str = "_nodes_history"
+
+    def __init__(
+        self, cache_vars: Union[List[str], None] = None, cache_path: str = "./hamilton-cache"
+    ):
+        """Initialize the cache
+
+        :param cache_vars: List of nodes for which to store/load results. Passing None will use the cache
+        for all nodes. Default is None.
+        :param cache_path: File path to the cache. The file name doesn't need an extension.
+        """
+        self.cache_vars = cache_vars if cache_vars else []
+        self.cache_path = cache_path
+        self.cache = shelve.open(self.cache_path)
+        self.nodes_history: Dict[str, List[str]] = self.cache.get(
+            key=CacheAdapter.nodes_history_key, default=dict()
+        )
+        self.used_nodes_hash: Dict[str, str] = dict()
+
+    def run_before_graph_execution(self, *, graph: HamiltonGraph, **kwargs):
+        """Set `cache_vars` to all nodes if received None during `__init__`"""
+        if self.cache_vars == []:
+            self.cache_vars = [n.name for n in graph.nodes]
+
+    def run_to_execute_node(
+        self, *, node_name: str, node_callable: Any, node_kwargs: Dict[str, Any], **kwargs
+    ):
+        """Create cache key based on node callable hash (equiv. to HamiltonNode.version) and
+        the node inputs (`node_kwargs`).If key in cache (cache hit), load result; else (cache miss),
+        compute the node and append node name to `used_nodes_hash`.
+
+        Note:
+            - the callable hash is stored  in `used_nodes_hash` because it's required to create the
+            key in `run_after_node_execution` and the callable won't be accessible to recompute it
+        """
+        if node_name not in self.cache_vars:
```

Comment:
For afterwards -- we may want to make this a set? Odds are high this is not a slow operation, but it could get `O(n**2)` which we don't want.